src/core/finalize.ts

```diff
@@ -43,163 +47,280 @@ export function processResult(result: any, scope: ImmerScope) {
 	rootScope: ImmerScope,
-	parentState: undefined | ImmerState,
-	targetObject: any,
-	prop: string | number,
-	childValue: any,
-	rootPath?: PatchPath,
-	targetIsSet?: boolean
+	parent: ImmerState,
+	child: ImmerState,
+	key: string | number | symbol
 ) {
-	if (childValue == null) {
-		return
-	}
-
-	if (typeof childValue !== "object" && !targetIsSet) {
-		return
-	}
-	const childIsFrozen = isFrozen(childValue)
-	if (childIsFrozen && !targetIsSet) {
-		return
-	}
-
-	if (process.env.NODE_ENV !== "production" && childValue === targetObject)
-		die(5)
-	if (isDraft(childValue)) {
-		const path =
-			rootPath &&
-			parentState &&
-			parentState!.type_ !== ArchType.Set && // Set objects are atomic since they have no keys.
-			!has((parentState as Exclude<ImmerState, SetState>).assigned_!, prop) // Skip deep patches for assigned keys.
-				? rootPath!.concat(prop)
-				: undefined
-		// Drafts owned by `scope` are finalized here.
-		const res = finalize(rootScope, childValue, path)
-		set(targetObject, prop, res)
-		// Drafts from another scope must prevented to be frozen
-		// if we got a draft back from finalize, we're in a nested produce and shouldn't freeze
-		if (isDraft(res)) {
-			rootScope.canAutoFreeze_ = false
-		} else return
-	} else if (targetIsSet) {
-		targetObject.add(childValue)
-	}
-	// Search new objects for unfinalized drafts. Frozen objects should never contain drafts.
-	if (isDraftable(childValue) && !childIsFrozen) {
-		if (!rootScope.immer_.autoFreeze_ && rootScope.unfinalizedDrafts_ < 1) {
-			// optimization: if an object is not a draft, and we don't have to
-			// deepfreeze everything, and we are sure that no drafts are left in the remaining object
-			// cause we saw and finalized all drafts already; we can stop visiting the rest of the tree.
-			// This benefits especially adding large data tree's without further processing.
-			// See add-data.js perf test
+	parent.callbacks_.push(function childCleanup(patches, inversePatches) {
+		const state: ImmerState = child
+
+		// Can only continue if this is a draft owned by this scope
+		if (!state || !isSameScope(state, rootScope)) {
 			return
 		}
-		if (
-			parentState &&
-			parentState.base_ &&
-			parentState.base_[prop] === childValue &&
-			childIsFrozen
-		) {
-			// Object is unchanged from base - no need to process further
-			return
+
+		// Handle potential set value finalization first
+		fixPotentialSetContents(state)
+
+		const finalizedValue = getFinalValue(state)
+
+		// Update all locations in the parent that referenced this draft
+		updateDraftInParent(parent, state.draft_ ?? state, finalizedValue, key)
+
+		generatePatchesAndFinalize(state, patches, inversePatches)
+	})
+}
+
+function generatePatchesAndFinalize(
+	state: ImmerState,
+	patches?: Patch[],
+	inversePatches?: Patch[]
+) {
+	const shouldFinalize =
+		state.modified_ &&
+		!state.finalized_ &&
+		(state.type_ === ArchType.Set ||
+			(state.assigned_ && state.assigned_.size > 0))
```

Comment:
if `assigned_` is always initialized (as currently the case), the left branch is always true? Otherwise `state.assigned_?.size > 0` as a NIT?